### DISCUSSION TITLE
Add text on .jpg images

### DISCUSSION THREAD

**Discussion by satimis** [ImageMagick/ImageMagick#5907] [General]
Hi all,

What will the command line adding following text on center top of .jpg images in the folder
```
The Establishment Ceremony
Reynold Industrial Co., Ltd.
```
(2 lines)
Color
Line 1 - red
Line 2 - yellow

The names of .jpg images are not in order

Regards

**Comment by fmw42** [comment]
Just use -annotate twice with different +Y values for +X+Y in the command line so that the second line is below the first line. Give that a try.

**Reply by satimis** [reply]
`Just use -annotate twice with different +Y values for +X+Y in the command line so that the second line is below the first line. Give that a try.`
Ran following command;
$ convert image_01.jpg -gravity north -font arial -pointsize 80 -fill red -annotate +30 "The Establishment Ceremony" image_01_annotate.jpg

Even single -annotate can't work.  Please refers to image attached

Regards
![20220619_anne01_baby_edited_text_V_annotate](https://user-images.githubusercontent.com/30752517/209463200-730991e2-048e-47f6-ab30-0a66724d77e5.jpg)

**Reply by satimis** [reply]
Hi Fred,

Now I got it done, running following command line
`$ convert image_01.jpg -gravity north -font arial -pointsize 80 -fill red -annotate +0+30 "The Establishment Ceremony" -fill yellow -annotate +0+110 "Reynold Industrial Co. Ltd." image_01.jpg_annotate.jpg`

X must be +0
repeat -fill twice
repeat -annotate twice.

Please refers to attached image

Now please advise how to batch process the images.jpg?  They are not in serial numbers in the folder

Thanks

Regards
![20220619_anne01_baby_edited_text_V_annotate](https://user-images.githubusercontent.com/30752517/209464229-cd5094ee-0c05-430d-9c8d-d5331bcdd3bc.jpg)

**Comment by fmw42** [comment]
If it is the same text, then use mogrify in place of convert.  See https://imagemagick.org/Usage/basics/#mogrify.  For IM 7 you would use magick mogrify.

**Reply by satimis** [reply]
`If it is the same text, then use mogrify in place of convert. See https://imagemagick.org/Usage/basics/#mogrify. For IM 7 you would use magick mogrify.`
$ mogrify *.jpg -gravity north -font arial -pointsize 80 -fill red -annotate +0+30 "The Establishment Ceremony" -fill yellow -annotate +0+110 "Reynold Industrial Co. Ltd." annotate.jpg
Convert also works

But if I expect to retain the name of the original file, only adding _annotate.jpg what shall I add on the command line?

Thanks
Regards

**Reply by satimis** [reply]
Furthermore;
To underline the text lines how to do it?
Regards

**Reply by fmw42** [reply]
`mogrify -format jpg -path path_to_output_directory -gravity north -font arial -pointsize 80 -fill red -annotate +0+30 "The Establishment Ceremony" -fill yellow -annotate +0+110 "Reynold Industrial Co. Ltd." *.jpg`

The *.jpg at the end is both the input and output file.  Mogrify reads each (JPG) file in the current directory, processes it and writes it back to the input directory overwriting the input file. If you do not want the overwriting, then use the -path option to specify an existing empty directory where you want the output file to be placed.  The -format jpg defines the suffix for the output if you want it to be differenent (such as png).

**Reply by fmw42** [reply]
The only way I know for -annotate is to use an underlined font.  You can do it in Pango, see https://imagemagick.org/Usage/text/#pango_markup or in -draw with decorate, see https://imagemagick.org/script/magick-vector-graphics.php

**Reply by satimis** [reply]
Hi Fred,
`
The *.jpg at the end is both the input and output file. Mogrify reads each (JPG) file in the current directory, processes it and writes it back to the input directory overwriting the input file. If you do not want the overwriting, then use the -path option to specify an existing empty directory where you want the output file to be placed. The -format jpg defines the suffix for the output if you want it to be differenent (such as png).`
Sorry, I don't follow your advice.

Following 2 original images are on the folder:
rey_ind_opening_01_edited_20220729_042302.jpg
rey_ind_opening_02_edited_20220729_042333.jpg

1)
If I expect retain the output images as;
rey_ind_opening_01_edited_20220729_042302_annotate.jpg
rey_ind_opening_02_edited_20220729_042333_annotate.jpg
in the same folder, what will be the command?

2)
If I expect the output images on a new folder, what will be the command?  Shall I create the new folder in advance before running the command ?

Thanks

Regards

**Reply by satimis** [reply]
Hi Fred.
`The only way I know for -annotate is to use an underlined font. You can do it in Pango, see https://imagemagick.org/Usage/text/#pango_markup or in -draw with decorate, see https://imagemagick.org/script/magick-vector-graphics.php`
Your advice noted.  Thanks

I found Arial Black installed on Ubuntu 22.04 on Libre Office Writer.
What will be the command?
-font "Arial Black" didn't work
-font arialblack didn't work

Thanks

**Reply by fmw42** [reply]
My understanding is that you cannot rename the output. It will always be the same as the input. Find some other tool to rename files afterwards. Alternately, you can write a script FOR loop over each file in the input folder and use convert. But that is OS dependent.

Yes, you have to create the new folder ahead of time

For new fonts, use the path to the font file with suffix such as path_to/arial.ttf

What is your IM version and platform/OS?

**Reply by fmw42** [reply]
Arial Black is not an underline font to my knowledge!

**Reply by fmw42** [reply]
In mogrify, you can have -format annotate.jpg, but it will add a period before the annotate in the file name, e.g. input is file.jpg and output will be file.annotate.jpg.  But you cannot use an underscore.

**Reply by satimis** [reply]
`Arial Black is not an underline font to my knowledge!`
No !
I expect using it as BOLD

What will be the command ?

Thanks

**Reply by satimis** [reply]
`In mogrify, you can have -format annotate.jpg, but it will add a period before the annotate in the file name, e.g. input is file.jpg and output will be file.annotate.jpg. But you cannot use an underscore.`
It doesn't matter just making the output files different from their original files.

I have tried here to do it without success.  Please advise.  Thanks

**Reply by fmw42** [reply]
use -font path_to/arial_black.ttf assuming the font is a ttf font and you have to fill in the "path_to" with the actual path on your system to where you have put the arial_black.ttf font and assuming it is named arial_black.  Change it if the name is different.

In mogrify, try -format annotate.jpg

**Reply by satimis** [reply]
Hi Fred,
`
Yes, you have to create the new folder ahead of time`
Noted and thanks

`For new fonts, use the path to the font file with suffix such as path_to/arial.ttf`
No, not displayed as arial.ttf
They are displayed as;
Arial Black
C059
Comic Sans MS
etc.

What will be the command"

**Reply by fmw42** [reply]
Where did you put your font?  What is the path to them?  Put the full path to the font file with suffix in double quotes if they have a space in the name.

**Reply by satimis** [reply]
`Where did you put your font? What is the path to them? Put the full path to the font file with suffix in double quotes if they have a space in the name.`
My problem is;
I couldn't find the folder/directory on Ubuntu 22.04 for storing/installing fonts.  I have spent almost the whole day googling it without result, including posting it on Ubuntu.org forum without reply.

**Reply by fmw42** [reply]
try

`sudo find /usr -name font`

should tell you where your font directory is.

When you downloaded that arial bold font, where did you put it?

**Reply by satimis** [reply]
```
try
sudo find /usr -name font
should tell you where your font directory is.
```
$ sudo find /usr -name font
/usr/share/groff/1.22.4/font
/usr/share/mime/font

$ ls /usr/share/groff/1.22.4/font/
devascii  devlatin1  devps  devutf8
$ ls /usr/share/mime/font/
collection.xml  otf.xml  ttf.xml  woff2.xml  woff.xml
`

When you downloaded that arial bold font, where did you put it?`
I haven't installed fonts on Ubuntu 22.04

Regards

**Reply by fmw42** [reply]
I don't see any fonts.  Perhaps just download some ttf font from the Internet and put it anywhere you want on your system. Then reference it in your command by its path to the file with the filename and suffix. There are lots of free font sites from which you can download.

You never told me what your IM version is?  What do you get from 

`magick -version`

or if that fails

`convert -version`

**Reply by satimis** [reply]
$ magick -version
```
Command 'magick' not found, did you mean:
  command 'magic' from deb magic (8.3.105+ds.1-1.1)
  command 'magics' from deb magics++ (4.10.1-1)
Try: sudo apt install <deb name>
```
$ convert -version
```
Version: ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP(4.5) 
Delegates (built-in): bzlib djvu fftw fontconfig freetype heic jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib

```

**Comment by fmw42** [comment]
You can download and run the script at https://www.imagemagick.org/Usage/scripts/imagick_type_gen to collect all your fonts on your system so that you can use them by name (assuming you have fonts in normal places). Download it to your home directory for example. Then

```
cd to your home directory
chmod u+x imagick_type_gen
imagick_type_gen > type-satimis.xml

```

(or your can use any name as type-any_name.xml)

Then you can do 

`cat type-satimis.xml`

to list your fonts names, paths and filenames. Once you have that file created, you can use fonts in your command by name rather than the path to the filename.suffix

**Reply by satimis** [reply]
Performed following steps;

Download the script and save it /home/sattims as imagick_type_gen

On Terminal
$ cd /home/satimis

$ chmod u+x imagick_type_gen 
$ ls -al imagick_type_gen 
-rwxrw-r-- 1 satimis satimis 17044 Dec 27 13:59 imagick_type_gen

$ imagick_type_gen > type-satimis.xml
imagick_type_gen: command not found

Also tried "sudo chmod u+x imagick_type_gen" with the same result

Regards

**Reply by fmw42** [reply]
Sorry, it is a perl script. So try

```
chmod u+x imagick_type_gen.pl
imagick_type_gen.pl  > type-satimis.xml
```

**Reply by satimis** [reply]
```
chmod u+x imagick_type_gen.pl
imagick_type_gen.pl  > type-satimis.xml
```
On Terminal performed following steps;

$ mv imagick_type_gen imagick_type_gen.pl
$ chmod u+x imagick_type_gen.pl
$ ls -al imagick_type_gen.pl 
-rwxrw-r-- 1 satimis satimis 17044 Dec 27 13:59 imagick_type_gen.pl

$ imagick_type_gen.pl > type-satimis.xml
imagick_type_gen.pl: command not found

Still the same

**Reply by satimis** [reply]
Hi Fred,

I found following link on Internet;

ImageMagick and imagick_type_gen don't recognize added fonts
https://stackoverflow.com/questions/30085275/imagemagick-and-imagick-type-gen-dont-recognize-added-fonts

Regards

**Reply by fmw42** [reply]
Sorry, I am not a Linux expert. But the script was written by a Linux expert. So it should work if your systems is set up compatibly.  In any case, you can still use any downloaded font, by use -font path_to_your_font/fontname.suffix.  The suffix generally is ttf.  Download ttf fonts and put them in any directory you want.

**Reply by satimis** [reply]
Hi Fred,

I got it done.  It is a little bid complicate.
I ran following steps on an Ubuntu 22.04 VM of Oracle VirtualBox
On Terminal
$ apt policy perl
```
perl:
  Installed: 5.34.0-3ubuntu1.1
  Candidate: 5.34.0-3ubuntu1.1
  Version table:
 *** 5.34.0-3ubuntu1.1 500
        500 http://hk.archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages
        500 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages
        100 /var/lib/dpkg/status
     5.34.0-3ubuntu1 500
        500 http://hk.archive.ubuntu.com/ubuntu jammy/main amd64 Packages
        
```
Make sure perl installed        

$ touch test.pl        
$ sudo nano test.pl 
copy following line on the file
#!/usr/bin/perl

and copy the download file on it

$ chmod +x test.pl

$ ./test.pl  > type-satimis.xml
```
sh: 1: locate: not found
sh: 1: locate: not found
sh: 1: locate: not found
sh: 1: locate: not found
```

$ locate bash_completion.sh
```
Command 'locate' not found, but can be installed with:
sudo apt install plocate
```

$ sudo apt install plocate

$ ./test.pl > type-satimis.xml
Now it works without complaint

type-satimis.xml is big file.

```
<?xml version="1.0"?>
<typemap>

  <type
     format="ttf"
     name="Tuffy"
     fullname="Tuffy Regular"
     family="Tuffy"
     glyphs="/etc/apt/ImageMagick-7.1.0-37/PerlMagick/demo/Generic.ttf"
     />

  <type
     format="ttf"
     name="Tuffy"
     fullname="Tuffy Regular"
     family="Tuffy"
     glyphs="/etc/apt/ImageMagick-7.1.0-37/PerlMagick/t/Generic.ttf"
     />

  <type
     format="ttf"
     name="Tuffy"
     fullname="Tuffy Regular"
     family="Tuffy"
     glyphs="/etc/apt/ImageMagick-7.1.0-37/PerlMagick/t/ttf/input.ttf"
     />

  <type
     format="ttf"
     name="TwemojiMozilla"
     fullname="Twemoji Mozilla"
     family="Twemoji Mozilla"
     glyphs="/snap/firefox/2154/usr/lib/firefox/fonts/TwemojiMozilla.ttf"
     />
```

What shall I do the next ?  Thanks

Is it the full path of the font folder/directory ?
`/etc/apt/ImageMagick-7.1.0-37/PerlMagick/demo/Generic.ttf`

Thanks

**Reply by satimis** [reply]
Hi Fred,

Just got a reply from Ubuntu Forum
`The output of 'convert -list font' or 'identify -list font' might be helpful in this case ...`

The output looks similar to the content of type-satimis.xml

Regards

**Reply by fmw42** [reply]
Correct, but for IM 7, 

magick -list font

will now show your list of fonts that Imagemagick knows about. It needs that file you just created. But I still don't understand why you use test.pl.  The script I sent was not named that.  What are you doing?

**Comment by fmw42** [comment]
What is test.pl?  How is that connected to the script I linked for you?

Your font files (name.ttf) are in different directories. See the paths for the glyphs.  But you can use the full name now I believe. Be sure to enclose in quotes if there is a space in the name. Or try using the name.  I do not recall which is correct.  But you can use the path to the .ttf file or one of the names with -font. Imagemagick finds the fonts by name from the file you created. So just go ahead and try one in your command.

**Reply by satimis** [reply]
Hi Fred

For example
```
.....
convert -gravity north -font arial -pointsize 80 -fill red -annotate +0+30 "The Establishment Ceremony" -fill yellow -annotate +0+110 "Reynold Industrial Co. Ltd." ${name}_annotate.jpg
```

Full path of font folder;
```
...
-font "/etc/apt/ImageMagick-7.1.0-37/PerlMagick/demo/Generic.ttf" ....
```

Is it correct ?  Thanks

**Reply by fmw42** [reply]
Yes, if you actually have a font named arial, that is correct.  The second example with the path is correct also, but I do not know what the Generic.ttf font looks like. So you have to test both.

**Comment by satimis** [comment]
$ convert 20220619_anne01_baby_edited_text_V.jpg -gravity north -font "/etc/apt/ImageMagick-7.1.0-37/PerlMagick/demo/Generic.ttf" -pointsize 80 -fill red -annotate +0+30 "The Establishment Ceremony" -fill yellow -annotate +0+110 "Reynold Industrial Co. Ltd." anne01_annotate.jpg
`convert-im6.q16: unable to read font `/etc/apt/ImageMagick-7.1.0-37/PerlMagick/demo/Generic.ttf' @ warning/annotate.c/RenderType/949.
The path is not correct ?

Regards

`

**Reply by fmw42** [reply]
Try

`find /usr -name Generic.ttf`

What path does that give you?

If that fails, try

`sudo find / -name Generic.ttf`

**Reply by fmw42** [reply]
Send me your full list from 

`convert -list font`

**Reply by satimis** [reply]
Hi Fred,

Just got a reply from Ubuntu Forum

On Terminal
```
$ identify -list font | grep arial
    glyphs: /usr/share/fonts/truetype/msttcorefonts/arialbi.ttf
    glyphs: /usr/share/fonts/truetype/msttcorefonts/ariali.ttf
```
$ convert 20220619_anne01_baby_edited_text_V.jpg -gravity north -font "/usr/share/fonts/truetype/msttcorefonts/arialbi.ttf" -pointsize 80 -fill red -annotate +0+30 "The Establishment Ceremony" -fill yellow -annotate +0+110 "Reynold Industrial Co. Ltd." anne01_annotate.jpg

It works.  Can I change 'Arial" to another font?

**Reply by satimis** [reply]
```
Send me your full list from
convert -list font
```
$ convert -list font > font_list.txt
[font_list.txt](https://github.com/ImageMagick/ImageMagick/files/10312086/font_list.txt)

**Reply by satimis** [reply]
```
Try

find /usr -name Generic.ttf
```
No output

```
If that fails, try
sudo find / -name Generic.ttf
```
$ sudo find / -name Generic.ttf
```
find: ‘/run/user/1000/doc’: Permission denied
find: ‘/run/user/1000/gvfs’: Permission denied
```

**Reply by fmw42** [reply]
Any font in /usr/share/fonts/truetype... should be fine either as the name (Font:) or the path (glyphs).

```
  Font: Georgia-Bold
    family: Georgia
    style: Normal
    stretch: Normal
    weight: 700
    glyphs: /usr/share/fonts/truetype/msttcorefonts/georgiab.ttf

```
So you could use `-font Georgia-Bold` or `-font /usr/share/fonts/truetype/msttcorefonts/georgiab.ttf`

Give them a try

**Reply by satimis** [reply]
`convert -list font`
How to run above command to find truetype fonts

**Reply by fmw42** [reply]
`convert -list font | grep "ttf"`

**Reply by satimis** [reply]
$ convert -list font | grep "ttf" > ttf_list.txt

ttf_list.txt is attached here

Regards
[ttf_list.txt](https://github.com/ImageMagick/ImageMagick/files/10312558/ttf_list.txt)

**Reply by fmw42** [reply]
Any of those should be fine

**Comment by fmw42** [comment]
Sure, but the using the path listed under the glyphs in your listing should work, especially ones from /usr. Perhaps you do not have permissions to access the ones in /etc. I just do not know enough about Linux.   But the truetype ones should be OK.

Fred

**Reply by satimis** [reply]
```
So you could use -font Georgia-Bold or -font /usr/share/fonts/truetype/msttcorefonts/georgiab.ttf

Give them a try
```
Both work.  Thanks

**Reply by satimis** [reply]
Hi Fred,
1)
Could you please explain follows;

1a)
name=$(convert "$img" -format "%t" info:)
What is the function of "%t" ?

1b)
i=$((i+1))
?
2)
Can I erase the text alreay on the image ?

Thanks

Regards

**Reply by fmw42** [reply]
1) That syntax returns a number of different information (called string formats or escapes for each letter in the %_ format). The %t means the filename without the suffix. See https://imagemagick.org/script/escape.php

2) i=$((i+1) is just incrementing the counter variable i.  It just adds 1 to it 

3) No, you cannot erase.  The pixels are permanently changed by adding text. There is no "memory" of the original once you write text to the image. You can write over it with the same text in the very same location using the -fill color as whatever you have for the background. Or you can just draw a rectangle filled with the background color over the text area.

**Reply by satimis** [reply]
Hi Fred,
Your advice noted and thanks

`convert "$img" -gravity northwest -font arial-bold -pointsize 90 -fill yellow -annotate +100+100 "AR$ii" ${name}_annotate.jpg`

If I expect the name of the output image having the same number at the end to replace "annotate"

e.g.
image-1.jpg -> image-1_AA01.jpg
image-2.jpg -> image-2_AA02.jpg
image-2.jpg -> image-3_AA03.jpg
etc

How to change the line?  Thanks

Regards

**Reply by fmw42** [reply]
Try

`convert "$img" -set filename:f "%t" -gravity northwest -font arial-bold -pointsize 90 -fill yellow -annotate +100+100 "AR$ii" "%[filename:f]_annotate.jpg"`

-set filename:f "%t" sets the filename as from the input image without the suffix
%[filename:f] uses it as a variable in the output

The double quotes are important

**Reply by satimis** [reply]
Sorry, it can't work.

Performed following steps;
i=1
for img in *.jpg; do
name=$(convert "$img" -format "%t" info:)
ii=$(printf "%02d" "$I")
convert "$img" -set filename:f "%t" -gravity northwest -font arial-bold -pointsize 90 -fill yellow -annotate +100+100 "AR$ii" "%[filename:f]_annotate.jpg"
i=$((i+1))
done

The numbers are not attached to the end of the output files as
e.g.
image_01_AR01.jpg
image_01_AR02.jpg
image_01_AR03.jpg
etc.

The output images only display;
image_01_annotate.jpg
image_02_annotate.jpg
image_03_annotate.jpg
etc.

Regards

**Reply by fmw42** [reply]
So do you just want

```
i=1
for img in *.jpg; do
name=$(convert "$img" -format "%t" info:)
ii=$(printf "%02d" "$i")
convert "$img" -set filename:f "%t" -gravity northwest -font arial-bold -pointsize 90 -fill yellow -annotate +100+100 "AR$ii" "%[filename:f]_AR${ii}_annotate.jpg"
i=$((i+1))
done
```

**Reply by satimis** [reply]
```
$ i=1
$ for img in *.jpg; do
> name=$(convert "$img" -format "%t" info:)
> ii=printf "%02d" "$i"
> convert "$img" -set filename:f "%t" -gravity northwest -font arial-bold -pointsize 90 -fill yellow -annotate +100+100 "AR$ii" "%[filename:f]_AR${ii}_annotate.jpg"
> i=$((i+1))
> done
```
bash: fg: %02d: no such job
bash: fg: %02d: no such job
bash: fg: %02d: no such job
bash: fg: %02d: no such job
bash: fg: %02d: no such job

```
$ i=1
$ for img in *.jpg; do
> name=$(convert "$img" -format "%t" info:)
> ii=`printf "%02d" "$i"`
> convert "$img" -set filename:f "%t" -gravity northwest -font arial-bold -pointsize 90 -fill yellow -annotate +100+100 "AR$ii" "%[filename:f]_AR${ii}_annotate.jpg"
> i=$((i+1))
> done
```
It works.

If creating a new folder for the output images where shall I put it in the command line?
`> convert "$img" -set filename:f "%t" -gravity northwest -font arial-bold -pointsize 90 -fill yellow -annotate +100+100 "AR$ii" ~/path/to/new_folder/"%[filename:f]_AR${ii}_annotate.jpg"`
???

Thanks

**Reply by fmw42** [reply]
Move your quote to the beginning and end of the output:

`convert "$img" -set filename:f "%t" -gravity northwest -font arial-bold -pointsize 90 -fill yellow -annotate +100+100 "AR$ii" "~/path/to/new_folder/%[filename:f]_AR${ii}_annotate.jpg"`

**Reply by satimis** [reply]
Hi Fred,
Your advice work.  Thank.
I have tried my version without moving " (double quote) to the front of / but failed.
If the file names are random e.g.;
4p01_20220619.jpg
rey_baby_01_20220619.jpg 
rey_child_01_20220619_text.jpg
etc.
How to change 
`%[filename:f]_AR${ii}_annotate.jpg`
to retain the original file names before AR${ii}_annotation ?

Could you please point me where can I find relevant documents learning ImageMagick. Google search found me many suggestion.  Thank.

Furthermore what is the command adding text in the center of the image ?

Regards

**Reply by fmw42** [reply]
For new users, see

http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=9620
http://www.imagemagick.org/script/command-line-processing.php
https://imagemagick.org/script/command-line-options.php
http://www.imagemagick.org/Usage/reference.html
http://www.imagemagick.org/Usage/
https://github.com/ImageMagick/usage-markdownhttps://imagemagick.org/script/porting.php#cli
https://legacy.imagemagick.org/discourse-server/index.php

**Reply by satimis** [reply]
Thanks for the URLs
